src/common/wal/src/lib.rs

```diff
@@ -30,3 +35,43 @@ pub enum TopicSelectorType {
     #[default]
     RoundRobin,
 }
+
+pub async fn resolve_to_ipv4(broker_endpoint: &str) -> Result<String> {
+    net::lookup_host(broker_endpoint)
+        .await
+        .context(ResolveEndpointSnafu { broker_endpoint })?
+        // only IPv4 addresses are valid
+        .find_map(|addr| addr.is_ipv4().then_some(addr.to_string()))
+        .context(EndpointIPV4NotFoundSnafu { broker_endpoint })
+}
+
+#[cfg(test)]
+mod tests {
+    use std::assert_matches::assert_matches;
+
+    use super::*;
+    use crate::error::Error;
+
+    // test for resolve_broker_endpoint
+    #[tokio::test]
+    async fn test_valid_host() {
+        let host = "localhost:9092";
+        let got = resolve_to_ipv4(host).await;
+        assert_eq!(got.unwrap(), "127.0.0.1:9092");
+    }
+
+    #[tokio::test]
+    async fn test_valid_host_ipv6() {
+        // the host is valid, it is an IPv6 address, but we only accept IPv4 addresses
```

Comment:
Out of this thread, I wonder if we should support IPv6 sooner or later ...

cc @fengjiachun @sunng87